src/vmm/src/device_manager/mmio.rs

```diff
@@ -548,4 +556,34 @@ mod tests {
             .get_device(DeviceType::Virtio(type_id), &id)
             .is_none());
     }
+
+    #[test]
+    fn test_slot_irq_allocation() {
+        let start_addr1 = GuestAddress(0x0);
+        let start_addr2 = GuestAddress(0x1000);
+        let guest_mem =
+            GuestMemoryMmap::from_ranges(&[(start_addr1, 0x1000), (start_addr2, 0x1000)]).unwrap();
+        let mut vm = builder::setup_kvm_vm(&guest_mem, false).unwrap();
+
+        #[cfg(target_arch = "x86_64")]
+        assert!(builder::setup_interrupt_controller(&mut vm).is_ok());
+        #[cfg(target_arch = "aarch64")]
+        assert!(builder::setup_interrupt_controller(&mut vm, 1).is_ok());
+
+        let mut device_manager =
+            MMIODeviceManager::new(&mut 0xd000_0000, (arch::IRQ_BASE, arch::IRQ_MAX));
+        let _addr = device_manager.allocate_new_slot(0);
+        assert_eq!(device_manager.irq, arch::IRQ_BASE);
+        let _addr = device_manager.allocate_new_slot(1);
+        assert_eq!(device_manager.irq, arch::IRQ_BASE + 1);
+        assert_eq!(
+            format!(
+                "{}",
+                device_manager
+                    .allocate_new_slot(arch::IRQ_MAX + 1)
```

Comment:
> Here I assume you want to test the smallest number of irqs that can no longer be allocated. In this case, this should be replaced with:
> `.allocate_new_slot(arch::IRQ_MAX - arch::IRQ_BASE + 1)`

Ta,  Done. 

> 
> Moreover, I would test some more corner case scenarios, e.g.: all the irqs are taken, but you try to allocate a new slot with 0 irqs (this case should be successful), you have only one irq available but you try to reserve two etc.

Given the agreement to open a follow up PR to discuss resource allocation, I would not add additional tests here.  Since the current allocation implementation might be fated, I would be concerned that further tests are just churn.  Discussion as part of a specific PR, and then additional tests if required, sounds better to me.

Reply:
I might be too rigid here, but I don't really agree with you on that:
- we don't know when that PR will be opened, merged etc.; it looks like a re-design is involved there and I suppose it will take time;
- you are changing the logic here for a pretty important function; since I expect we want to keep the new irqs allocation logic (i.e. let the device request as many irqs as it wants) in the other design as well, these tests won't be a waste of time at all, they will probably need some `allocate_new_slot()` -> `allocate_irqs()` change in them; moreover you will make the re-design implementer more aware of what needs to be tested.

Reply:
Hi @lauralt,  apologies for the delay getting back to you.  I've attempted to add the two tests you've highlighted, if you had a moment to have a look...

Reply:
It's great, thanks! Overall, the PR looks good, I guess we should update now the boottime image with the init wrapper, and then we will let you know when it's done.